### PR TITLE
Type stats follow renaming

### PR DESCRIPTION
**Pull request by marcotc**
**What does this PR do?**

Makes typing stats comparison follow renames and line moves for RBS declarations and `steep:ignore` comments.

**Motivation:**

Without this change, type reports currently show file renames as "introduced + cleared".

[Example](https://github.com/DataDog/dd-trace-rb/pull/5804#issuecomment-4514741951):
```
Untyped methods (+3 -3)
Partially typed methods (+5 -5)
```

**Change log entry**

None.

**Additional Notes:**

Tests were added for `.github/scripts/typing_stats_compare.rb`, for both old and new behavior.
To ensure test run, I added it to the same Raketask as our custom Rubocop cops (`spec:custom_cop`), but renamed the task to `spec:tooling`.
These tests do not run on `spec:main`.

**How to test the change?**

`bundle exec rake spec:tooling`

### CODE REVIEW

File: .github/scripts/typing_stats_compute.rb [hunk 1]
```diff
@@ -39,40 +39,63 @@
...
+  rbs_buffer = ::RBS::Buffer.new(name: path, content: source)
   comments.each do |comment|
     ignore = ::Steep::AST::Ignore.parse(comment, rbs_buffer)
     next if ignore.nil? || ignore.is_a?(::Steep::AST::Ignore::IgnoreEnd)
 
-    result << {
-      path: path.to_s,
-      line: ignore.line
-    }
+    ignored_source = source_lines[ignore.line - 1]&.strip
```
[the remark this answers is not in the document]
Comment (vpellan) [reply .github/scripts/typing_stats_compute.rb]: I agree, now that we're not comparing with line numbers anymore but the content of the ignored lines, we should probably use everything that is between the steep:ignore:start and steep:ignore:end (which would also correctly report any change between start and end). But maybe this should be for another PR

File: .github/scripts/typing_stats_compute.rb [hunk 2]
```diff
@@ -39,40 +39,63 @@
...
+def ast_traversal(declarations, result = {}, context = [])
   result[:methods] ||= []
   result[:others] ||= []
   declarations.each do |declaration|
     case declaration
     when ::RBS::AST::Declarations::Module,
          ::RBS::AST::Declarations::Class,
          ::RBS::AST::Declarations::Interface
-      ast_traversal(declaration.members, result)
+      ast_traversal(declaration.members, result, context + [declaration.name.to_s])
```
Comment (vpellan) [inline .github/scripts/typing_stats_compute.rb]: When you rename a file, you'll usually also rename the corresponding Module, including in the .rbs file, which means the `context` will be different. I don't see any operation on `context` in `typing_stats_compare.rb` so I assume it will report them as two different declaration and not fix the issue you're trying to solve here. We could probably normalize the file name, and when comparing the `context`, check every module name against their corresponding renamed one (although there's most likely some edge cases but it would work most of the time)

I'd also add a test for that (the current tests do not rename modules)

**Review by vpellan — CHANGES_REQUESTED** [review]
Awesome work, thank you! I never really had time to go back to this tooling, but this should prevent a lot of noise due to renaming files or adding/removing lines. I still requested a change because as-is, this PR might not work as expected (see my comment in AST traversal)

**Review by Strech — APPROVED** [review]
